In [6]:
import pandas as pd

In [7]:
df0 = pd.read_csv('Titanic-Dataset.csv')

In [8]:
df0.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [9]:
df0.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [10]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [11]:
df0.shape

(891, 12)

In [12]:
df0.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [13]:
# conseguir datos nulos
df0.isnull().sum().sort_values(ascending=False)

Cabin          687
Age            177
Embarked         2
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
SibSp            0
Parch            0
Ticket           0
Fare             0
dtype: int64

In [14]:
# pasajeros por clase
df0.groupby('Pclass').count()

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               216       216   216  216  186    216    216     216   216   
2               184       184   184  184  173    184    184     184   184   
3               491       491   491  491  355    491    491     491   491   

        Cabin  Embarked  
Pclass                   
1         176       214  
2          16       184  
3          12       491

In [15]:
df0['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [16]:
# agrupando para ver de que puertos salieron cada clase
df0.groupby(['Pclass', 'Embarked']).count()

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  \
Pclass Embarked                                                                
1      C                  85        85    85   85   74     85     85      85   
       Q                   2         2     2    2    2      2      2       2   
       S                 127       127   127  127  108    127    127     127   
2      C                  17        17    17   17   15     17     17      17   
       Q                   3         3     3    3    2      3      3       3   
       S                 164       164   164  164  156    164    164     164   
3      C                  66        66    66   66   41     66     66      66   
       Q                  72        72    72   72   24     72     72      72   
       S                 353       353   353  353  290    353    353     353   

                 Fare  Cabin  
Pclass Embarked               
1      C           85     66  
       Q            2      2  
       S          127    106  
2      C           17      2  
       Q            3      1  
       S          164     13  
3      C           66      1  
       Q           72      1  
       S          353     10

In [17]:
127*100/216

58.7962962962963

In [18]:
# completar datos faltantes en una columna
df1 = df0.copy()

In [19]:
df_embarked = df1[df1['Embarked'].isnull()]

In [20]:
# vemos quienes no sabemos de que puerto salieron
df_embarked

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

In [21]:
df1.groupby(['Pclass', 'Embarked'])['Fare'].describe()

count        mean        std      min      25%      50%  \
Pclass Embarked                                                            
1      C          85.0  104.718529  99.093935  26.5500  49.5000  78.2667   
       Q           2.0   90.000000   0.000000  90.0000  90.0000  90.0000   
       S         127.0   70.364862  58.811278   0.0000  29.2500  52.0000   
2      C          17.0   25.358335  11.345067  12.0000  13.8625  24.0000   
       Q           3.0   12.350000   0.000000  12.3500  12.3500  12.3500   
       S         164.0   20.327439  13.630741   0.0000  13.0000  13.5000   
3      C          66.0   11.214083   4.871528   4.0125   7.2292   7.8958   
       Q          72.0   11.183393   6.721677   6.7500   7.7500   7.7500   
       S         353.0   14.644083  13.276609   0.0000   7.8542   8.0500   

                        75%       max  
Pclass Embarked                        
1      C         110.883300  512.3292  
       Q          90.000000   90.0000  
       S          83.475000  263.0000  
2      C          37.004200   41.5792  
       Q          12.350000   12.3500  
       S          26.000000   73.5000  
3      C          14.458300   22.3583  
       Q          10.218725   29.1250  
       S          16.100000   69.5500

In [22]:
# rellenar los valores faltantes de Embarked con S
df1["Embarked"].fillna('S', inplace=True)

C:\Users\Desiree\AppData\Local\Temp\ipykernel_8712\1328197415.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df1["Embarked"].fillna('S', inplace=True)


In [23]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [24]:
df1['Embarked'].unique()

array(['S', 'C', 'Q'], dtype=object)

In [25]:
df1.groupby("Pclass").count()

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               216       216   216  216  186    216    216     216   216   
2               184       184   184  184  173    184    184     184   184   
3               491       491   491  491  355    491    491     491   491   

        Cabin  Embarked  
Pclass                   
1         176       216  
2          16       184  
3          12       491

### Como vemos, los valores de Embarked que antes eran 889 de 891. Ahora estan completos

In [26]:
column_list = df1.columns.to_list()

In [27]:
# se extrajo la columna correspondiente a las cabinas
column_list.pop(-2)

'Cabin'

In [28]:
column_list

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Embarked']

In [29]:
# creamos un nuevo data frame con las columnas que vamos a utilizar
df2 = df1[column_list]

In [30]:
df2.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S

In [31]:
# edades faltantes por clase y por sexo
df_age_c1 = df2[(df2['Pclass'] == 1) & df2['Age'].isnull()]

In [32]:
# guardo edades faltantes en la clase 1 sexo femenino
df_age_c1_female = df_age_c1[df_age_c1['Sex'] == 'female']

In [33]:
df_age_c1_female

PassengerId  Survived  Pclass  \
31            32         1       1   
166          167         1       1   
256          257         1       1   
306          307         1       1   
334          335         1       1   
375          376         1       1   
457          458         1       1   
669          670         1       1   
849          850         1       1   

                                                  Name     Sex  Age  SibSp  \
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female  NaN      1   
166             Chibnall, Mrs. (Edith Martha Bowerman)  female  NaN      0   
256                     Thorne, Mrs. Gertrude Maybelle  female  NaN      0   
306                            Fleming, Miss. Margaret  female  NaN      0   
334  Frauenthal, Mrs. Henry William (Clara Heinshei...  female  NaN      1   
375              Meyer, Mrs. Edgar Joseph (Leila Saks)  female  NaN      1   
457                  Kenyon, Mrs. Frederick R (Marion)  female  NaN      1   
669  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  female  NaN      1   
849       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female  NaN      1   

     Parch    Ticket      Fare Embarked  
31       0  PC 17569  146.5208        C  
166      1    113505   55.0000        S  
256      0  PC 17585   79.2000        C  
306      0     17421  110.8833        C  
334      0  PC 17611  133.6500        S  
375      0  PC 17604   82.1708        C  
457      0     17464   51.8625        S  
669      0     19996   52.0000        S  
849      0     17453   89.1042        C

In [34]:
df_age_c1_male = df_age_c1[df_age_c1['Sex'] == 'male']

### Edades clase 1

In [35]:
# edades faltantes por clase 1
df_age_c1 = df2[(df2['Pclass'] == 1) & df2['Age'].isnull()]

# guardo las edades faltantes en la clase 1 sexo female
df_age_c1_female = df_age_c1[df_age_c1['Sex'] == 'female']

#guardo las edades faltantes en la clase 2 sexo male
df_age_c1_male = df_age_c1[df_age_c1['Sex'] == 'male']

In [37]:
print(f'Edades faltantes clase 1, sexo femenino: {df_age_c1_female.shape[0]}')
print(f'Edades faltantes clase 1, sexo masculino: {df_age_c1_male.shape[0]}')

Edades faltantes clase 1, sexo femenino: 9
Edades faltantes clase 1, sexo masculino: 21


### Edades clase 2 

In [38]:
# edades faltantes por clase 2 
df_age_c2 = df2[(df2['Pclass'] == 2) & df2['Age'].isnull()]

# guardo las edades faltantes en la clase 2 sexo female
df_age_c2_female = df_age_c2[df_age_c2['Sex'] == 'female']

#guardo las edades faltantes en la clase 2 sexo male
df_age_c2_male = df_age_c2[df_age_c2['Sex'] == 'male']

In [40]:
print(f'Edades faltantes clase 2, sexo femenino: {df_age_c2_female.shape[0]}')
print(f'Edades faltantes clase 2, sexo masculino: {df_age_c2_male.shape[0]}')

Edades faltantes clase 2, sexo femenino: 2
Edades faltantes clase 2, sexo masculino: 9


### Edades clase 3

In [41]:
# edades faltantes por clase 3 
df_age_c3 = df2[(df2['Pclass'] == 3) & df2['Age'].isnull()]

# guardo las edades faltantes en la clase 3 sexo female
df_age_c3_female = df_age_c3[df_age_c3['Sex'] == 'female']

#guardo las edades faltantes en la clase 3 sexo male
df_age_c3_male = df_age_c3[df_age_c3['Sex'] == 'male']

In [42]:
print(f'Edades faltantes clase 3, sexo femenino: {df_age_c3_female.shape[0]}')
print(f'Edades faltantes clase 3, sexo masculino: {df_age_c3_male.shape[0]}')

Edades faltantes clase 3, sexo femenino: 42
Edades faltantes clase 3, sexo masculino: 94


In [43]:
# sacar el promedio de edades de cada sexo por clase

In [45]:
df3 = df2.groupby(['Pclass', 'Sex'])['Age'].describe()
df3

count       mean        std   min     25%   50%    75%   max
Pclass Sex                                                                 
1      female   85.0  34.611765  13.612052  2.00  23.000  35.0  44.00  63.0
       male    101.0  41.281386  15.139570  0.92  30.000  40.0  51.00  80.0
2      female   74.0  28.722973  12.872702  2.00  22.250  28.0  36.00  57.0
       male     99.0  30.740707  14.793894  0.67  23.000  30.0  36.75  70.0
3      female  102.0  21.750000  12.729964  0.75  14.125  21.5  29.75  63.0
       male    253.0  26.507589  12.159514  0.42  20.000  25.0  33.00  74.0

In [46]:
# acceso mas directo a un dato del data frame
c1_age_female = df3.loc[(1, 'female'), 'mean']
c1_age_male = df3.loc[(1, 'male'), 'mean']

c2_age_female = df3.loc[(2, 'female'), 'mean']
c2_age_male = df3.loc[(2, 'male'), 'mean']

c3_age_female = df3.loc[(3, 'female'), 'mean']
c3_age_male = df3.loc[(3, 'male'), 'mean']

### Completar las edades faltantes por clase por sexo

In [47]:
df_age_c1_female['Age'] = df_age_c1_female['Age'].fillna(round(c1_age_female,1))

df_age_c2_female['Age'] = df_age_c2_female['Age'].fillna(round(c2_age_female,1))

df_age_c3_female['Age'] = df_age_c3_female['Age'].fillna(round(c3_age_female,1))


C:\Users\Desiree\AppData\Local\Temp\ipykernel_8712\3704238516.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_age_c1_female['Age'] = df_age_c1_female['Age'].fillna(round(c1_age_female,1))
C:\Users\Desiree\AppData\Local\Temp\ipykernel_8712\3704238516.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_age_c2_female['Age'] = df_age_c2_female['Age'].fillna(round(c2_age_female,1))
C:\Users\Desiree\AppData\Local\Temp\ipykernel_8712\3704238516.py:5: SettingWithCopyWarning: 
A value is trying to be set o

In [48]:
df_age_c3_female

PassengerId  Survived  Pclass  \
19            20         1       3   
28            29         1       3   
32            33         1       3   
47            48         1       3   
82            83         1       3   
109          110         1       3   
128          129         1       3   
140          141         0       3   
180          181         0       3   
186          187         1       3   
198          199         1       3   
229          230         0       3   
235          236         0       3   
240          241         0       3   
241          242         1       3   
264          265         0       3   
274          275         1       3   
300          301         1       3   
330          331         1       3   
347          348         1       3   
358          359         1       3   
359          360         1       3   
367          368         1       3   
368          369         1       3   
409          410         0       3   
415          416         0       3   
431          432         1       3   
485          486         0       3   
502          503         0       3   
533          534         1       3   
564          565         0       3   
573          574         1       3   
578          579         0       3   
593          594         0       3   
612          613         1       3   
653          654         1       3   
680          681         0       3   
697          698         1       3   
727          728         1       3   
792          793         0       3   
863          864         0       3   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
19                             Masselmani, Mrs. Fatima  female  21.8      0   
28                       O'Dwyer, Miss. Ellen "Nellie"  female  21.8      0   
32                            Glynn, Miss. Mary Agatha  female  21.8      0   
47                           O'Driscoll, Miss. Bridget  female  21.8      0   
82                      McDermott, Miss. Brigdet Delia  female  21.8      0   
109                                Moran, Miss. Bertha  female  21.8      1   
128                                  Peter, Miss. Anna  female  21.8      1   
140                      Boulos, Mrs. Joseph (Sultana)  female  21.8      0   
180                       Sage, Miss. Constance Gladys  female  21.8      8   
186    O'Brien, Mrs. Thomas (Johanna "Hannah" Godfrey)  female  21.8      1   
198                   Madigan, Miss. Margaret "Maggie"  female  21.8      0   
229                            Lefebre, Miss. Mathilde  female  21.8      3   
235                       Harknett, Miss. Alice Phoebe  female  21.8      0   
240                              Zabour, Miss. Thamine  female  21.8      1   
241                     Murphy, Miss. Katherine "Kate"  female  21.8      1   
264                                 Henry, Miss. Delia  female  21.8      0   
274                         Healy, Miss. Hanora "Nora"  female  21.8      0   
300           Kelly, Miss. Anna Katherine "Annie Kate"  female  21.8      0   
330                                 McCoy, Miss. Agnes  female  21.8      2   
347          Davison, Mrs. Thomas Henry (Mary E Finck)  female  21.8      1   
358                               McGovern, Miss. Mary  female  21.8      0   
359                  Mockler, Miss. Helen Mary "Ellie"  female  21.8      0   
367                     Moussa, Mrs. (Mantoura Boulos)  female  21.8      0   
368                                Jermyn, Miss. Annie  female  21.8      0   
409                                 Lefebre, Miss. Ida  female  21.8      3   
415            Meek, Mrs. Thomas (Annie Louise Rowley)  female  21.8      0   
431  Thorneycroft, Mrs. Percival (Florence Kate White)  female  21.8      1   
485                             Lefebre, Miss. Jeannie  female  21.8      3   
502                     O'Sullivan, Miss. Bridget Mary  female  21.8      0   


In [49]:
df_age_c1_male['Age'] = df_age_c1_male['Age'].fillna(round(c1_age_male,1))

df_age_c2_male['Age'] = df_age_c2_male['Age'].fillna(round(c2_age_male,1))

df_age_c3_male['Age'] = df_age_c3_male['Age'].fillna(round(c3_age_male,1))

C:\Users\Desiree\AppData\Local\Temp\ipykernel_8712\3806215816.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_age_c1_male['Age'] = df_age_c1_male['Age'].fillna(round(c1_age_male,1))
C:\Users\Desiree\AppData\Local\Temp\ipykernel_8712\3806215816.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_age_c2_male['Age'] = df_age_c2_male['Age'].fillna(round(c2_age_male,1))
C:\Users\Desiree\AppData\Local\Temp\ipykernel_8712\3806215816.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of 

In [50]:
df_age_c2_male

PassengerId  Survived  Pclass                              Name   Sex  \
17            18         1       2      Williams, Mr. Charles Eugene  male   
181          182         0       2                  Pernot, Mr. Rene  male   
277          278         0       2       Parkes, Mr. Francis "Frank"  male   
413          414         0       2    Cunningham, Mr. Alfred Fleming  male   
466          467         0       2             Campbell, Mr. William  male   
481          482         0       2  Frost, Mr. Anthony Wood "Archie"  male   
547          548         1       2        Padro y Manent, Mr. Julian  male   
674          675         0       2        Watson, Mr. Ennis Hastings  male   
732          733         0       2              Knight, Mr. Robert J  male   

      Age  SibSp  Parch         Ticket     Fare Embarked  
17   30.7      0      0         244373  13.0000        S  
181  30.7      0      0  SC/PARIS 2131  15.0500        C  
277  30.7      0      0         239853   0.0000        S  
413  30.7      0      0         239853   0.0000        S  
466  30.7      0      0         239853   0.0000        S  
481  30.7      0      0         239854   0.0000        S  
547  30.7      0      0  SC/PARIS 2146  13.8625        C  
674  30.7      0      0         239856   0.0000        S  
732  30.7      0      0         239855   0.0000        S